Content.Server/Fluids/EntitySystems/MoppingSystem.cs

```diff
@@ -0,0 +1,363 @@
+
+                    sfx = component.TransferSound;
+                }
+            }
+            else return; // if the target did not have a Refillable Component
+        }
+        else if (currentVolume <= 0) // tool is dry
+        {
+            if (TryComp<DrainableSolutionComponent>(target, out var drainable))
+            {
+                // These return conditions will abort BEFORE the do_after is called:
+                if (!_solutionSystem.TryGetDrainableSolution(target, out var drainableSolution))
+                {
+                    return;
+                }
+                else if (drainableSolution.CurrentVolume <= 0) // target container is empty (liquid source)
+                {
+                    msg = "mopping-system-target-container-empty";
+                    user.PopupMessage(user, Loc.GetString(msg, ("target", target))); // play message now because we are returning.
+                    return;
+                }
+                else
+                {
+                    // Determine transferAmount
+                    transferAmount = FixedPoint2.Min(availableVolume * 0.5, drainableSolution.CurrentVolume); // Let's transfer up to to half the tool's available capacity to the tool.
+
+                    donor = target; // the drainable container's Uid
+                    donorSolutionName = drainable.Solution;
+
+                    acceptor  = used; // the mop/tool Uid
+                    acceptorSolutionName = "absorbed"; // by definition on AbsorbentComponent
+
+                    // Set delay/popup/sound if nondefault. Popup and sound will only play on a successful doAfter.
+                    // default delay is fine for this case.
+                    msg = "mopping-system-drainable-success";
+                    sfx = component.TransferSound;
+                }
+            } else return; // if the target did not have a Drainable Component
+        }
+        else return;
+
+        var doAfterArgs = new DoAfterEventArgs(user, delay, target: target)
+        {
+            BreakOnUserMove = true,
+            BreakOnStun = true,
+            BreakOnDamage = true,
+            MovementThreshold = 0.2f,
+            BroadcastCancelledEvent = new TransferCancelledEvent()
+            {
+                Target = target,
+                Component = component // (the AbsorbentComponent)
+            },
+            BroadcastFinishedEvent = new TransferCompleteEvent()
+            {
+                User = user,
+                Tool = used,
+                Target = target,
+                Donor = donor,
+                Acceptor = acceptor,
+                Component = component,
+                DonorSolutionName = donorSolutionName,
+                AcceptorSolutionName = acceptorSolutionName,
+                Message = msg,
+                Sound = sfx,
+                TransferAmount = transferAmount
```

Comment:
this really makes me think that you should probably just figure out this stuff in the doafter and not beforehand

Reply:
Yeah, this is one of the things I went back and forth on.

I saw option 1 as being cleaner/less expensive, but I could just as easily implement option 2 if you prefer. You might also spot a better alternative option 3 that my untrained eye missed. Happy to do it whichever way is most maintainable.

Option 2 looks worse on a flowchart, but tbh it might be better because the portion before the doAfter would become really lightweight, postponing the heavy lifting until the doAfter actually succeeds.

# Option 1
![image](https://user-images.githubusercontent.com/97707302/155855732-5c9dc617-bd9f-4e02-a49f-ba1da177991c.png)

# Option 2
![image](https://user-images.githubusercontent.com/97707302/155855949-a30a38ff-1d4a-48b3-b74e-163f39729962.png)

Reply:
(Obviously if we go with Option 2 or something else, I'll tweak the flowchart properly in the PowerPoint so it doesn't look like hot garbage)

Reply:
Option 2 does not really make sense, because right now all 3 interactions use the same exact `OnTransferComplete` and `OnTransferCancelled` path (you don't branch anywhere in them depending on which interaction was chosen before the DoAfter). Like you said, you are basically going to repeat the checking logic which is silly.

Reply:
Thanks Acruid! I think I'll keep it the way I have it (Option 1) to keep the logic cleaner. Those five args are passed here because they depend on which of the three actions the mopper is doing.

The one thing I'll try for control flow is separate the logic for the three actions into their own reusable functions, as noted above. We'll see if that makes it any cleaner.